middleware.go

```diff
@@ -106,6 +117,98 @@ func createMiddleware(mw TykMiddleware) func(http.Handler) http.Handler {
 	}
 }
 
+// loopbackCORS sends an OPTIONS request via loopback to the upstream.
+// Given: CORS is not handled by the gateway
+//   and OptionsPassthrough is set to true
+// When: an Origin header is present on the request
+// Then: it performs an OPTIONS request via loopback and writes the upstream CORS headers to the error response
+func loopbackCORS(w http.ResponseWriter, r *http.Request, mw TykMiddleware) {
+	log := log.WithField("prefix", "middleware.loopbackCORS")
+
+	if mw.Base().Spec.CORS.Enable {
+		log.Debug("gateway handling CORS - returning")
+		return
+	}
+
+	if !mw.Base().Spec.CORS.OptionsPassthrough {
+		log.Debug("options passthrough is false - returning")
+		return
+	}
+
+	if r.Header.Get("Origin") == "" {
+		log.Debug("no origin header set in request so not a browser - returning")
+		return
+	}
+
+	// we don't need the request body for a CORS request
+	scheme := "http"
+	if config.Global.HttpServerOptions.UseSSL {
+		scheme += "s"
+	}
+
+	u := fmt.Sprintf("%s://localhost:%d/%s", scheme, config.Global.ListenPort, r.URL.Path)
+	corsRequest, err := http.NewRequest(http.MethodOptions, u, nil)
+
+	// https://www.w3.org/TR/2014/REC-cors-20140116/#cross-origin-request-with-preflight-0
+
+	corsRequest.Header.Set("Content-Type", r.Header.Get("Content-Type"))
+	corsRequest.Header.Set(request.Origin, r.Header.Get("Origin"))
+	corsRequest.Header.Set(request.AccessControlRequestMethod, r.Method)
+
+	// If author request headers is not empty include an Access-Control-Request-Headers header with as header field
+	// value a comma-separated list of the header field names from author request headers in lexicographical order,
+	// each converted to ASCII lowercase (even when one or more are a simple header).
+	var requestHeaders []string
+	for key := range r.Header {
+		requestHeaders = append(requestHeaders, strings.ToLower(key))
+	}
+	sort.Strings(requestHeaders)
+
+	corsRequest.Header.Set(request.AccessControlRequestHeaders, strings.Join(requestHeaders, ", "))
+
+	// TODO: Exclude Credentials
+	// The term user credentials for the purposes of this specification means cookies, HTTP authentication, and
+	// client-side SSL certificates that would be sent based on the user agent's previous interactions with the
+	// origin. Specifically it does not refer to proxy authentication or the Origin header. [COOKIES]
+
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	dump, err := httputil.DumpRequest(corsRequest, true)
+	if err != nil {
+		http.Error(w, fmt.Sprint(err), http.StatusInternalServerError)
+		return
+	}
+
+	log.Debugf("loopback request dump\n%s", string(dump))
+
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: config.Global.HttpServerOptions.SSLInsecureSkipVerify},
+	}
+
+	corsClient := &http.Client{Transport: tr}
+	corsRes, err := corsClient.Do(corsRequest)
```

Comment:
this needs to have some kind of circuit breaker or at least time out for client to avoid waiting for upstream when it is down (if we don't it could cause bump in gateway latency)

Reply:
So the circuit breaker I am not sure is required. because we are calling Tyk itself but this time, as a pre-flight request.

Reply:
oh, I though we were going to send OPTIONS to upstream directly. it seems to me that calling GW itself again is an extra step (but it should work). I'd add timeout to request anyways

Reply:
That's why it's called loopback. because the gw calls it's own API. The reasoning for this is because the OPTIONS request needs to go through the exact same MW chain as the original pre-flight request would have done before so that we get the same request to the same upstream endpoint.

e.g. there may be MW in place which re-writes the target.

Reply:
ok, got it. according to original pre-flight request - some browsers don't send pre-flight requests for all ajax calls, so original OPTIONS one might never happened. I don't think it would cause any issues in this case, just wanted to mention it.

Reply:
I am still not sure about this loopback round trip business, as looks like all we want to do is just to hit the code placed a couple of lines above here: https://github.com/TykTechnologies/tyk/pull/1550/files#diff-4c166b743d95629a9ab37eb6fe3df22aR86

why not just to prepare OPTIONS request like we do anyways and response writer (we can probably use httptest recoder) and then call `h.ServeHTTP(optionsRecorder, optionsRequest)` to make it go through all underlaying middle-wares down to upstream?